nixos/modules/config/users-groups.nix

```diff
@@ -598,7 +598,35 @@ in {
           Neither the root account nor any wheel user has a password or SSH authorized key.
           You must set one to prevent being locked out of your system.'';
       }
-    ];
+    ]
+    ++ flip mapAttrsToList cfg.users (name: user:
+      { assertion =
+        # This regex matches a subset of Modular Crypto Format (MCF)[1]
+        # informal standard.  Since this depends largely on the OS or the
+        # specific implementation of crypt(3) we only support the (sane)
+        # schemes implemented by glibc and BSDs. In particular the original
+        # DES hash is excluded since, having no structure, it would validate
+        # common mistakes like typing the plaintext password.
+        #
+        # [1]: https://en.wikipedia.org/wiki/Crypt_(C)
+        let
+          sep = "\\$";
+          base64 = "[a-zA-Z0-9./]+";
+          id = "[a-z0-9-]+";
+          value = "[a-zA-Z0-9/+.-]+";
+          options = "${id}(=${value})?(,${id}=${value})*";
+          scheme  = "${id}(${sep}${options})?";
+          content = "${base64}${sep}${base64}";
+          mcf = "^${sep}${scheme}${sep}${content}$";
+        in
+          user.hashedPassword != null
+          -> builtins.match mcf user.hashedPassword != null;
+        message = ''
+          The password hash of user "${name}" is not valid. You must set a valid hash or
```

Comment:
How do we distinguish between “not valid” and “we didn’t think about this implementation in our assertion check?”

Reply:
Currently we don't. Since this format is not specified in any standard I think the he only reasonable way is to match the hashes supported by glibc (which I believe is the only usable libc in NIxOS) and a few other common ones.

To avoid the rejecting valid hashes we could either make this only a warning or add a mechanism to let users override the check, maybe something with prioities.